File: Backend/smarttransithub/src/main/java/com/backend/smarttransithub/services/FcmService.java
package com.backend.smarttransithub.services;


import org.springframework.stereotype.Service;

import com.google.firebase.messaging.FirebaseMessaging;
import com.google.firebase.messaging.Message;
import com.google.firebase.messaging.Notification;



@Service

public class FcmService {
    public void sendNotification(String token, String title, String body)
    {
        if(token==null || token.isBlank())
        {
            System.out.println("Token is null or blank. Cannot send notification.");
            return;
        }
        //Build the notification in Notification Class.
        Notification notification=Notification.builder()
            .setTitle(title)
            .setBody(body)
            .build();

        //Message Creation.
        Message message=Message.builder()
            .setToken(token)
            .setNotification(notification)
            .build();


        FirebaseMessaging.getInstance().sendAsync(message);
    }
}

File: NotificationService/Service/FcmService.cs
using FirebaseAdmin;
using FirebaseAdmin.Messaging;
using Google.Apis.Auth.OAuth2;
using NotificationService.Models;
using System.Net;

namespace NotificationService.Service
{
    public class FcmService : IFcmService
    {

        public FcmService(IConfiguration configuration)
        {
            if (FirebaseApp.DefaultInstance == null)
            {
                string? path = configuration["Firebase:ServiceAccountPath"];

                if (string.IsNullOrWhiteSpace(path))
                    throw new InvalidOperationException(
                        "Firebase service account path is not configured.");

                if (!File.Exists(path))
                    throw new FileNotFoundException(
                        $"Firebase service account file not found: {path}");

                using var stream = File.OpenRead(path);

                var serviceAccount =
                    ServiceAccountCredential.FromServiceAccountData(stream);

                FirebaseApp.Create(new AppOptions
                {
                    Credential =
                        GoogleCredential.FromServiceAccountCredential(
                            serviceAccount)
                });
            }
        }

        public async Task<ApiResponse> SendNotificationAsync(NotificationRequest request)
        {
            if (string.IsNullOrWhiteSpace(request.Token))
            {
                return new ApiResponse
                {
                    Status = "Failure",
                    Message = "Invalid Device Token"
                };
            }

            var message = new Message
            {
                Token = request.Token,

                Notification = new Notification
                {
                    Title = request.Title,
                    Body = request.Body
                }
            };

            try
            {
                string response = await FirebaseMessaging.DefaultInstance.SendAsync(message);

                return new ApiResponse
                {
                    Status = "Success",
                    Message = response
                };
            }
            catch(Exception ex)
            {
                return new ApiResponse
                {
                    Status = "Failure",
                    Message = ex.Message
                };
            }
        }
    }
}
